### DISCUSSION TITLE
How to initialize a composable asynchronously

### DISCUSSION THREAD

**Discussion by lacorde** [nuxt/nuxt#18458] [Questions]
Let's say I need to initialize my `useTest` composable at app creation, with a value fetched from my API.
I cannot do the following or the fetch will trigger every time I call `useTest`.

```
export const useTest = async () => {
  const test = useState("test", () => '');
  test.value = $fetch('https://myapi.com');
  return test;
};
```

What is the right way to do this ? Somewhere else ? (in a plugin maybe)

**Comment by oritwoen** [comment]
What about this? Then fetch should theoretically only run if no data exists.

```ts
export const useTest = async () => {
  const test = useState("test", () => '');
  
  if (!test.value) {
    test.value = await $fetch('https://myapi.com');
  }
  
  return test;
};
```

**Reply by lacorde** [reply]
That's a pretty good approach if useTest is never to be "emptied". Thanks.

**Comment by ndragun92** [comment]
From what u have described it feels like the best approach would be a mix of store (Pinia) and a plugin.
Once the app is initialized plugin will be executed and there u can fetch data and store it in the store so you can use it globally. I do not know what is ur use case for that since context is missing

**Reply by lacorde** [reply]
This is my current approach, thanks. Would you recommend Pinia over composables for store ? why ?

**Reply by Baroshem** [reply]
I would say that just the plugin will be enough. No need for Pinia unless there are any special cases there. I am using a plugin to initialize Algolia in my Algolia module and later on, I have access to a generated client from the NuxtApp.

https://github.com/nuxt-modules/algolia/blob/main/src/runtime/plugin.ts
